### PR TITLE
ClientSocket Refactor

### PR DESCRIPTION
**Pull request by bwoebi**
PR for allowing comments

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,6 +1,7 @@
 The MIT License (MIT)
 
-Copyright (c) 2015 Bob Weinand.
+Copyright (c) 2015-2017 amphp
+Copyright (c) 2015 Bob Weinand
```
Comment (bwoebi) [inline LICENSE]: Just put it all under © amphp.

File: lib/Processor.php [hunk 2]
```diff
@@ -147,52 +143,54 @@ private function ready() {
         }
     }
 
-    public function connect() {
+    public function connect(): Promise {
         \assert(!$this->deferreds && !$this->socket, self::class."::connect() must not be called twice");
 
-        $this->deferreds[] = $deferred = new Deferred;
-        \Amp\Socket\connect($this->config->resolvedHost)->onResolve(function ($error, $socket) use ($deferred) {
-            if ($this->connectionState === self::CLOSED) {
```
Comment (bwoebi) [inline lib/Processor.php]: we get $this->connectionState === self::CLOSED here when the user calls closeSocket() before connection is established. This must be safe here.

File: lib/Processor.php [hunk 3]
```diff
@@ -841,29 +833,13 @@ public function initClosing() {
...
     }
 
-    private function compilePacket() {
-        do {
-            $pending = current($this->out);
-            unset($this->out[key($this->out)]);
-            if ($pending !== null || empty($this->out)) {
-                break;
-            }
-            $this->seqId = $this->compressionId = -1;
```
Comment (bwoebi) [inline lib/Processor.php]: the seqId/compressionId is not reset in the new code - that's likely causing bugs

File: lib/Processor.php [hunk 4]
```diff
@@ -890,78 +866,28 @@ private function compilePacket() {
...
-    private function compressPacket($packet) {
-        $packet = $this->uncompressedOut.$packet;
-
+    private function compressPacket(string $packet): string {
         if ($packet == "") {
             return "";
         }
 
         $len = strlen($packet);
-        while ($len < self::MAX_UNCOMPRESSED_BUFLEN && !empty($this->out)) {
```
Comment (bwoebi) [inline lib/Processor.php]: self::MAX_UNCOMPRESSED_BUFLEN is important to ensure that packets larger than this are properly split up - you've just straight ahead removed the whole logic for that.

**Comment by trowski** [comment]
@bwoebi I believe this is close to working… it's as though the connect promise is not being resolved. Define `MYSQL_DEBUG` and try example 2. Perhaps you'll be able to see where I'm going wrong.

**Comment by trowski** [comment]
@bwoebi With the latest commits all tests and examples work with the exception of `PoolTest::testVirtualConnection`, which just hangs. Any insight to what might be causing that?

**Comment by kelunik** [comment]
Build on 7.1+ segfault now.

**Comment by trowski** [comment]
@bwoebi The 7.1 build failing might be a phpdbg thing, as it passes locally with xdebug on 7.1.11 running the `stream_select` loop.

File: lib/Processor.php [hunk 5]
```diff
@@ -1193,26 +1171,23 @@ private function sendHandshake($inSSL = false) {
...
-                /* wait until main write watcher has written everything... */
-                if ($this->outBuflen > 0 || !empty($this->out)) {
+            \Amp\call(function () use ($payload) {
+                yield $this->write($payload);
+
+                $context = $this->config->ssl ?: new ClientTlsContext;
+                $context = $context->withPeerName($this->config->host);
+
+                return yield $this->socket->enableCrypto($context);
+            })->onResolve(function ($error) {
```
Comment (kelunik) [inline lib/Processor.php]: Why is that inside `onResolve` and not directly inside the coroutine?

Comment (trowski) [reply lib/Processor.php]: Oversight when refactoring. This onResolve function existed in the prior version at [line 1205-1210](https://github.com/amphp/mysql/pull/49/files/604fd53853a69dc133f48ee41a3d0098cbf38788#diff-b7c6c28a97dddfea08068fa97b76dafdL1205). I'll roll it into the coroutine.

**Pull request merged**